### ISSUE TITLE
Tab key indent on multiple-line selection is weird if any line does not match indent levels

### ISSUE THREAD

**Issue by ThiefMaster** [microsoft/vscode#104371]
Issue Type: <b>Bug</b>

I have a file with this content, indentation set to 2 spaces
```
foo
  bar
   foobar
```

If I select all lines and press TAB once, I would expect two spaces to be added in front of every single line:

```
  foo
    bar
     foobar
```

However, it seems like vscode is trying to be smart and indent each line to the next level of indentation (ie n*2 spaces):

```
  foo
    bar
    foobar
```

This is incredibly annoying. For example look what happens in my Jinja HTML template (with attributes being aligned below the start of the first one) with this:

```html
<a class="foobar"
   title="Test">
    {% trans %}Withdraw{% endtrans %}
</a>
```

After indenting it (4 space indentation here) I get this nonsense:

```html
    <a class="foobar"
    title="Test">
        {% trans %}Withdraw{% endtrans %}
    </a>
```

To me it feels like a bug, but in case people may actually like this it should at least be configurable...

VS Code version: Code - Insiders 1.48.0-insider (2277c8e2a3e1cc630a6397301ba54a1dccd8a60d, 2020-08-07T11:32:09.993Z)
OS version: Windows_NT x64 10.0.19041
Remote OS version: Linux x64 4.19.97-gentoo

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 7 2700X Eight-Core Processor          (16 x 3693)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off_ok<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.93GB (18.92GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|SSH: blackhole|
|OS|Linux x64 4.19.97-gentoo|
|CPUs|Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz (8 x 1453)|
|Memory (System)|15.53GB (0.86GB free)|
|VM|0%|
</details><details><summary>Extensions (33)</summary>

Extension|Author (truncated)|Version
---|---|---
systemd-unit-file|coo|1.0.6
udev-rules|dpk|1.0.0
viml|dun|0.1.7
tmux|mal|0.1.0
remote-containers|ms-|0.134.0
remote-ssh|ms-|0.51.0
remote-ssh-edit|ms-|0.51.0
remote-wsl|ms-|0.44.4
vscode-remote-extensionpack|ms-|0.20.0
jinjahtml|sam|0.13.0
gitconfig|sid|2.0.1
mako|tom|0.2.0
vscode-icons|vsc|10.2.0
overtype|ada|0.2.0
vscode-module-templates|asb|1.3.0
vscode-postgres|cko|1.1.11
quick-open-prefill-selection|col|0.0.1
vscode-eslint|dba|2.1.8
gitlens|eam|10.2.2
EditorConfig|Edi|0.15.1
prettier-vscode|esb|5.1.3
go|gol|0.16.1
search-editor-apply-changes|jak|0.1.1
vsc-python-indent|Kev|1.10.1
restructuredtext|lex|129.0.0
MagicPython|mag|1.1.0
vscode-github-actions|me-|3.0.1
vscode-docker|ms-|1.4.1
python|ms-|2020.8.93521-dev
vscode-pylance|ms-|2020.8.0
puppet-vscode|pup|0.28.0
vscode-yaml|red|0.9.1
vscode-stylelint|sty|0.84.0

(1 theme extensions excluded)

</details>

**Comment by OmeGak** [comment]
The behavior in action:

![ezgif-6-9feace463759](https://user-images.githubusercontent.com/716307/89789353-c512c700-db20-11ea-85db-f963bd5846ef.gif)

**Comment by alexdima** [comment]
You can control this and disable snapping to tab stops with `"editor.useTabStops": false`

**Comment by ThiefMaster** [comment]
@alexdima unfortunately this is not a solution. When I disable this setting I also lose the usual outdent-by-tabstop behavior of the backspace key.

Is there **any** case where the behavior from my original issue would be what people expect? I'd say no, so IMHO it's still a bug when you select multiple lines and indenting/dedenting them does not apply the same indent/dedent on every single line.

**Comment by OmeGak** [comment]
The bug, in my opinion, is that `"editor.useTabStops": true` is applied on every line independently. This is at least what is causing the inconsistent behavior. To prevent this behavior, what I would expect is the `"editor.useTabStops": true` being applied on the first line of the multiline selection and all the following lines being [de]indented the same number of spaces.

**Comment by alexdima** [comment]
I appreciate your opinion, but `"editor.useTabStops": true` is doing exactly what it was designed to do. If you want your code to be aligned at multiples of tab stops, then please continue using `"editor.useTabStops": true`. If you don't want your code to be aligned at multiples of tab stops, then please configure `"editor.useTabStops": false`. You cannot have the first line of a statement be aligned and the rest of the statement be not aligned, the editor logic regarding indenting and backspace is not semantic, it does not understand that the next lines contains the same statement, the editor just sees another line.

When `"editor.useTabStops": false` is configured, indenting will insert a tab stop worth of spaces or a tab character to the starts of all the selected lines. Unindenting will also do the same. This will keep the relative indentation of lines. Backspace will remove precisely one space to allow you to position things at non-multiples of tab stops, how else can you create that 3 spaces indent from 4 spaces.

**Comment by ThiefMaster** [comment]
But this whole issue is about cases where selecting multiple consecutive lines and then using tab or shift-tab applies **inconsistent** indentation changes to the different lines. Again, I don't think this is something you'd ever want. If you want to fix inconsistent indentation you either use a formatter or fix lines individually.

But the common case when selecting a block of text is that you want to indent/outdent that whole block! I think every other editor does this just fine. The current vscode behavior is inconsistent and right now cannot be disabled except by toggling a setting which also disables other, much more useful, functionality.

---

> how else can you create that 3 spaces indent from 4 spaces.

move cursor one left, and press "del". anyway, this sounds like a normal usecase for the `useTabStops` setting if you prefer to be able to use backspace for this.

> The best is to install a formatter that you like and understands these language specific things.

I agree, unfortunately there is none for jinja+html.

**Comment by alexdima** [comment]
@ThiefMaster 

By default `"editor.useTabStops"` is `true`. You do not agree with how that works. I get that and that is perfectly fine. Then please configure it to `"editor.useTabStops": false`.

The idea behind `"editor.useTabStops"` is that you want your code to align to tab stops or not. You declare that via a setting and then everything makes sure to keep your code aligned or not. We don't support out of the box that you have indenting not aligning to tab stops, but backspace aligning to tab stops. 

But that is not a problem! There are very many VS Code extensions that can get you covered.

e.g. Backspace++ https://marketplace.visualstudio.com/search?term=backspace&target=VSCode&category=All%20categories&sortBy=Relevance

 or maybe take a look at all the possible indentation extensions e.g. https://marketplace.visualstudio.com/search?term=indent&target=VSCode&category=All%20categories&sortBy=Relevance  maybe there is one that indents just the way you want it to.

VS Code is extensible, you are not forced to use only the builtin commands. The builtin commands are serving very many programming languages, and might not fit well with all programming languages, nor with all programming styles.

**Comment by ThiefMaster** [comment]
> The idea behind "editor.useTabStops" is that you want your code to align to tab stops or not.

I agree. But do you expect a simple indentation change to perform some kind of re-formatting on the code block you are re-indenting, and not just changing the indentation of the block as a whole?

I still believe this is something completely unrelated to whether one wants to use tabstops in all other cases or not!

**Comment by alexdima** [comment]
Yes, that is intentional. For example, I often copy code from something formatted with spaces (maybe stack overflow) and then I paste it somewhere where I use tabs and then I indent it once and most of the times the indent is fixed. Using `"editor.useTabStops"` makes it that indenting also replaces the wrongly indented lines with the correct indent. For example, here I copy some lines indented with spaces to an editor configured to indent with tabs. When I indent the lines, notice how a tab is inserted to each line, but also the 4 spaces are converted to tabs. So indenting snaps the lines to tab stops and fixes their indentation. This is what I want to happen, I don't want to end up in a mixed-indentation situation.

![Kapture 2020-09-04 at 16 19 32](https://user-images.githubusercontent.com/5047891/92249586-a0fd9800-eeca-11ea-848c-c40e82b902f7.gif)

**Comment by ThiefMaster** [comment]
Interesting usecase - but is that really an intended usage of the indent/dedent feature?

Wouldn't it be better to have a formatter for this, or an extension that fixes indentation style on paste? (scnr)

**Comment by alexdima** [comment]
> Interesting usecase - but is that really an intended usage of the indent/dedent feature?

Yes! That is why I implemented it like that! :) Definitely before we supported format on paste in the editor. I personally use it very often like that. If there is one thing I dislike more than mixed line endings, it is mixed whitespace in indentation. By the way: did you notice that VS Code is one of the only code editors that doesn't support mixed line endings. Mixed line endings have caused me so much pain in the past, so I built an editor where you cannot get mixed line endings :). But now some people want mixed line endings support :). What I'm saying is that sometimes the built-in things may be opinionated, and I am open to being wrong and changing them, but sometimes it is not so clear-cut, so that's why we added these options.

> Wouldn't it be better to have a formatter for this, or an extension that fixes indentation style on paste? 

Yes, formatters are awesome! But not all languages have formatters, and this behaviour is super helpful for the long tail of programming languages that don't have awesome language servers with good formatters available in VS Code. I sometimes even do `cmd+`a,`tab`, `shift+tab` to quickly "align" some code that I'm randomly working on.

I personally think you might be happy with `"editor.useTabStops": false` and `Backspace++`. Please, just give it a try. In fact Johannes did `Backspace++` before I had a chance to implement it in the editor's core, so my implementation is based on his extension.

**Comment by ThiefMaster** [comment]
Sure, I don't mind trying the extension. That might fix it for myself.

But I had a few cases where people in my team ended up having messed-up indentation in their PRs due to this strange behavior, and asking everyone to install an extension takes more effort than just trying to get it improved in the core :)

**Comment by alexdima** [comment]
I have to say, I often run into this problem too. Just the other day, I started using prettier in a couple of my projects. So far I have a good opinion of it and will try adding it to more repos. It turns out it was quite easy to add, and via `husky`, it is configured to run as a pre commit hook. See https://github.com/microsoft/monaco-editor-samples/commit/9dd7846d2542d925372b359827c5d96ee4e14ab0 . I found it liberating to just "give up" and delegate the formatting to an opinionated tool :). I could definitely add it to the CI, etc. But at least I don't need to think too much, I can just run `npx prettier --write .` if think something's off.
